In [0]:
from pyspark.sql.functions import col, lower, round, avg, count, sum, max, min, mean, countDistinct

spark.sql("USE CATALOG workspace")
spark.sql("USE SCHEMA default_cohort_db")


DataFrame[]

In [0]:
%sql
SELECT *
FROM ecom_orders


_fivetran_id order_date sales row_id customer_id order_id _fivetran_deleted _fivetran_synced 1FS46Z4iTYt4beJHHl8BA0cIM2Q= 2024-07-15 221.68 765 CUST157 ORD1764 false 2025-07-11T12:48:46.713Z +tZqj9WHsSAL7g9dSOVQLcHovwI= 2024-07-13 120.93 498 CUST183 ORD1497 false 2025-07-11T12:48:46.715Z LQyBZfO2hmqdWmzU7hdQNgJ3fzs= 2024-08-05 39.56 666 CUST093 ORD1665 false 2025-07-11T12:48:46.703Z UBdNvZSR0SOHC9VUTQbtC0GkYV4= 2024-06-06 10.83 123 CUST049 ORD1122 false 2025-07-11T12:48:46.689Z cyVYFlz3eQvnQoeHHygd2ik3i4U= 2024-02-04 148.01 439 CUST163 ORD1438 false 2025-07-11T12:48:46.713Z 2DMUzMuQB7tnNv2uwbN4Uz3wn14= 2024-03-21 73.74 60 CUST027 ORD1059 false 2025-07-11T12:48:46.679Z /mFYj/TBxfbhRGPPyCiIqJPOJzE= 2024-03-19 124.26 294 CUST112 ORD1293 false 2025-07-11T12:48:46.707Z G6dX4pZqI3f7YRY3uniBwLhgQoU= 2024-07-28 196.03 64 CUST028 ORD1063 false 2025-07-11T12:48:46.680Z FjGUfhS0oIfLsQzqnYpK231s9C4= 2024-06-15 21.1 469 CUST172 ORD1468 false 2025-07-11T12:48:46.714Z qEZexeWQvdPvdlvZcYR7WNaosYA= 2024-02-03 254.36 445 CUST165 ORD1444 false 2025-07-11T12:48:46.713Z RPL6xdMyypNh0IFZ9XDuRwV2yvs= 2024-12-01 275.32 955 CUST192 ORD1954 false 2025-07-11T12:48:46.716Z NTjYJNNnfeILdlBof1IeKxxEf60= 2024-08-01 114.35 51 CUST022 ORD1050 false 2025-07-11T12:48:46.677Z nUQaq+/sw3+FzffHdKL7aD9R2L4= 2024-07-03 161.65 969 CUST179 ORD1968 false 2025-07-11T12:48:46.715Z JWP86Ki/APBCZkRQNoONmb521Ls= 2024-10-30 294.95 852 CUST024 ORD1851 false 2025-07-11T12:48:46.678Z KqkM6G1bbtmpEAPby/FLlN3Ehg4= 2024-06-19 282.7 329 CUST125 ORD1328 false 2025-07-11T12:48:46.709Z 2TALB1jxnsQzbcqSUOpn+WjNIt8= 2024-10-11 200.46 888 CUST107 ORD1887 false 2025-07-11T12:48:46.706Z nt73GWxBDOHzhSX8YBe0dH3c0H8= 2024-07-13 55.3 78 CUST033 ORD1077 false 2025-07-11T12:48:46.682Z eLnvHHaUo1H0T4z9UmWHlHw/zHE= 2024-02-21 16.83 559 CUST186 ORD1558 false 2025-07-11T12:48:46.715Z 55NTxJXCqSGk8UVqtSYM490OHJw= 2024-03-04 190.24 807 CUST170 ORD1806 false 2025-07-11T12:48:46.714Z CaSIKwuOUk8aG+H+exHtye+o9yU= 2024-07-16 74.22 898 CUST142 ORD1897 false 2025-07-11T12:48:46.711Z HMUovrLwILxpGWAIYDSABibdXzo= 2024-10-25 18.33 653 CUST013 ORD1652 false 2025-07-11T12:48:46.671Z +1RF78CbI4a2tvSlTf6osq83MbQ= 2024-06-21 118.35 981 CUST032 ORD1980 false 2025-07-11T12:48:46.682Z qPxwPUs7PEg1PE7FHhQWWN9crhc= 2024-02-23 131.33 651 CUST105 ORD1650 false 2025-07-11T12:48:46.706Z 7qx0/rEBfZiEluQNdHoLkFJew8A= 2024-10-17 206.42 707 CUST135 ORD1706 false 2025-07-11T12:48:46.711Z NxVINlcvdbR7Y93hOornSN95eP4= 2024-05-20 22.25 827 CUST082 ORD1826 false 2025-07-11T12:48:46.701Z aiWtsDROSeUbGGHg1R5Inhqy2sM= 2024-03-21 89.58 71 CUST030 ORD1070 false 2025-07-11T12:48:46.681Z RU7FMobAL5bM2kJQqNI/vfJAZWs= 2024-04-18 72.28 360 CUST133 ORD1359 false 2025-07-11T12:48:46.710Z fL6ej2f4m3d9umUWmrJHiSelmIE= 2024-07-31 63.14 999 CUST036 ORD1998 false 2025-07-11T12:48:46.684Z 8/7SsVDtIQysiGZstNLMZipYAqM= 2024-11-12 199.59 929 CUST058 ORD1928 false 2025-07-11T12:48:46.694Z dfXG64+ZUNYoPtM0s25VFxfMT14= 2024-05-07 179.11 962 CUST025 ORD1961 false 2025-07-11T12:48:46.679Z 1xbi1baddiCXplDKaOKfvvvdvlE= 2024-07-26 251.35 2 CUST001 ORD1001 false 2025-07-11T12:48:46.662Z ZdoT9XoBywTZDdoEoYDQHlWdDyk= 2024-08-23 137.77 436 CUST161 ORD1435 false 2025-07-11T12:48:46.713Z hGGgFW0jr+YrHfG0mUZz4tap/Uo= 2024-06-07 156.85 412 CUST152 ORD1411 false 2025-07-11T12:48:46.712Z af4Qd1+vcBtEWHZNoVsn7AMj3P0= 2024-05-19 152.46 894 CUST052 ORD1893 false 2025-07-11T12:48:46.691Z tx1JO9kL7cQGBZ0k75PTy0hrhww= 2024-05-12 30.1 950 CUST185 ORD1949 false 2025-07-11T12:48:46.715Z vtY2jpMd9zNK+pANTNdNPywWrGk= 2024-11-18 163.42 904 CUST029 ORD1903 false 2025-07-11T12:48:46.681Z 82QlkpWC3TRuiBrcbnkZTM8F5uc= 2024-03-15 19.43 217 CUST081 ORD1216 false 2025-07-11T12:48:46.700Z uX5V3E/v8uDdgyJsKnQTNBS/ZvY= 2024-02-10 230.02 949 CUST147 ORD1948 false 2025-07-11T12:48:46.712Z MG4E03MeHp7NmM79avGdLUlAhzo= 2024-01-10 118.6 935 CUST101 ORD1934 false 2025-07-11T12:48:46.704Z zHT9hycCDR7MG9/KM7b4eTOW5Ww= 2024-04-14 197.33 410 CUST151 ORD1409 false 2025-07-11T12:48:46.712Z 5q4kcWODrIVH+

In [0]:
%sql
--Uniqueness Check. Check if order_id or row_id is unique.
SELECT row_id, COUNT(*) 
FROM ecom_orders
GROUP BY row_id
HAVING COUNT(*) > 1;


row_id COUNT(*)

In [0]:
%sql
----Uniqueness Check. Check if order_id or rowid is unique.

SELECT order_id, COUNT(*) AS cnt
FROM ecom_orders
GROUP BY order_id
HAVING COUNT(*) > 1;

order_id cnt

In [0]:
%sql
--Check for negative or zero sales
SELECT *
FROM ecom_orders
WHERE sales <= 0;

_fivetran_id order_date sales row_id customer_id order_id _fivetran_deleted _fivetran_synced

In [0]:
%sql
--Count Total Rows vs. Distinct Orders
SELECT COUNT(*) AS total_rows, 
COUNT(DISTINCT order_id) AS unique_orders
FROM ecom_orders;

total_rows unique_orders 1000 1000

# Step 1: Calculating the First Purchase Date
# 

In [0]:
# Step 1: First purchase date per customer
first_purchase_df = spark.sql("""
    SELECT
        customer_id,
        MIN(order_date) AS first_purchase_date -- earliest order date per customer
    FROM ecom_orders
    GROUP BY customer_id
""")

# Creating a temp view to use in next steps
first_purchase_df.createOrReplaceTempView("first_purchases")

display(first_purchase_df)


customer_id first_purchase_date CUST157 2024-01-03 CUST013 2024-04-30 CUST181 2024-01-03 CUST185 2024-03-04 CUST028 2024-06-20 CUST175 2024-05-05 CUST012 2024-02-15 CUST169 2024-02-05 CUST059 2024-02-21 CUST006 2024-01-18 CUST057 2024-01-16 CUST085 2024-03-16 CUST103 2024-02-16 CUST142 2024-03-28 CUST139 2024-04-18 CUST034 2024-01-14 CUST001 2024-04-12 CUST065 2024-01-03 CUST180 2024-03-06 CUST035 2024-05-31 CUST117 2024-06-13 CUST005 2024-03-16 CUST041 2024-01-25 CUST146 2024-01-21 CUST102 2024-01-14 CUST131 2024-01-20 CUST138 2024-03-29 CUST141 2024-02-17 CUST136 2024-05-18 CUST066 2024-02-18 CUST149 2024-01-31 CUST054 2024-01-09 CUST119 2024-01-31 CUST040 2024-02-06 CUST031 2024-03-14 CUST025 2024-02-28 CUST167 2024-01-06 CUST097 2024-03-04 CUST143 2024-01-26 CUST076 2024-01-14 CUST015 2024-04-15 CUST030 2024-01-01 CUST009 2024-06-01 CUST168 2024-05-02 CUST082 2024-04-20 CUST086 2024-02-08 CUST092 2024-01-09 CUST125 2024-05-21 CUST156 2024-04-23 CUST153 2024-02-06 CUST116 2024-01-09 CUST074 2024-02-01 CUST078 2024-04-08 CUST079 2024-04-29 CUST158 2024-03-20 CUST193 2024-05-11 CUST200 2024-01-29 CUST064 2024-06-08 CUST069 2024-01-21 CUST126 2024-01-31 CUST120 2024-06-21 CUST094 2024-04-03 CUST071 2024-02-21 CUST190 2024-02-06 CUST075 2024-04-30 CUST199 2024-03-19 CUST060 2024-01-10 CUST145 2024-02-25 CUST058 2024-06-20 CUST135 2024-03-25 CUST173 2024-02-19 CUST073 2024-03-26 CUST152 2024-02-14 CUST183 2024-04-23 CUST166 2024-02-03 CUST112 2024-03-03 CUST045 2024-01-29 CUST070 2024-04-04 CUST008 2024-01-24 CUST154 2024-04-21 CUST194 2024-04-10 CUST067 2024-02-09 CUST104 2024-01-04 CUST091 2024-01-19 CUST046 2024-01-28 CUST189 2024-01-20 CUST088 2024-03-05 CUST188 2024-01-06 CUST123 2024-03-13 CUST107 2024-04-22 CUST033 2024-02-13 CUST108 2024-01-10 CUST110 2024-02-08 CUST068 2024-02-02 CUST027 2024-03-21 CUST122 2024-02-15 CUST047 2024-05-08 CUST038 2024-01-01 CUST162 2024-01-07 CUST014 2024-05-20 CUST191 2024-04-03 CUST129 2024-05-10 CUST016 2024-01-22 CUST161 2024-02-13 CUST140 2024-02-25 CUST022 2024-06-05 CUST048 2024-04-20 CUST160 2024-02-25 CUST101 2024-01-10 CUST178 2024-06-18 CUST164 2024-05-15 CUST039 2024-01-16 CUST044 2024-06-02 CUST072 2024-01-02 CUST052 2024-04-02 CUST159 2024-03-08 CUST124 2024-03-04 CUST011 2024-01-07 CUST195 2024-04-08 CUST197 2024-03-15 CUST055 2024-02-03 CUST134 2024-04-15 CUST128 2024-06-22 CUST098 2024-01-20 CUST096 2024-02-29 CUST130 2024-02-16 CUST007 2024-02-02 CUST144 2024-03-25 CUST049 2024-03-12 CUST182 2024-01-25 CUST099 2024-01-30 CUST174 2024-01-25 CUST148 2024-03-23 CUST002 2024-04-11 CUST083 2024-01-23 CUST118 2024-05-01 CUST179 2024-01-18 CUST151 2024-03-03 CUST062 2024-01-22 CUST133 2024-03-11 CUST026 2024-02-14 CUST004 2024-01-31 CUST111 2024-02-22 CUST198 2024-01-21 CUST084 2024-04-24 CUST165 2024-02-03 CUST024 2024-03-07 CUST021 2024-03-11 CUST121 2024-01-18 CUST170 2024-01-21 CUST061 2024-01-27 CUST184 2024-02-25 CUST003 2024-01-14 CUST100 2024-03-20 CUST089 2024-01-03 CUST106 2024-03-27 CUST019 2024-01-20 CUST150 2024-01-24 CUST177 2024-02-20 CUST147 2024-02-10 CUST176 2024-04-05 CUST109 2024-01-13 CUST087 2024-01-25 CUST187 2024-03-13 CUST132 2024-05-03 CUST037 2024-05-14 CUST137 2024-01-02 CUST023 2024-02-15 CUST113 2024-02-04 CUST127 2024-03-07 CUST186 2024-02-03 CUST053 2024-02-13 CUST080 2024-04-13 CUST115 2024-01-06 CUST163 2024-02-04 CUST077 2024-01-26 CUST093 2024-02-18 CUST018 2024-04-01 CUST105 2024-02-23 CUST036 2024-05-02 CUST032 2024-02-02 CUST020 2024-01-13 CUST155 2024-05-06 CUST063 2024-01-15 CUST095 2024-01-07 CUST050 2024-02-06 CUST042 2024-04-07 CUST171 2024-05-05 CUST056 2024-04-29 CUST114 2024-02-21 CUST090 2024-01-15 CUST081 2024-03-15 CUST051 2024-02-15 CUST029 2024-02-07 CUST172 2024-01-25 CUST010 2024-03-16 CUST043 2024-02-06 CUST196 2024-03-02 CUST192 2024-01-02 CUST017 2024-01-11

# Step 2: Calculate the Second Purchase Date
# 

In [0]:
# Step 2: Second purchase date after first purchase
second_purchase_df = spark.sql("""
    SELECT
        o.customer_id,
        MIN(o.order_date) AS second_purchase_date -- order date after the first purchase
    FROM ecom_orders o
    INNER JOIN first_purchases fp
        ON o.customer_id = fp.customer_id
    WHERE o.order_date > fp.first_purchase_date -- only after first purchase
    GROUP BY o.customer_id
""")

# Save this for next join
second_purchase_df.createOrReplaceTempView("second_purchases")

display(second_purchase_df)

customer_id second_purchase_date CUST157 2024-02-04 CUST013 2024-06-23 CUST181 2024-02-11 CUST185 2024-03-26 CUST028 2024-07-28 CUST175 2024-05-30 CUST012 2024-03-04 CUST169 2024-03-14 CUST059 2024-03-12 CUST006 2024-03-31 CUST057 2024-05-01 CUST085 2024-05-02 CUST103 2024-02-24 CUST142 2024-05-30 CUST139 2024-04-25 CUST034 2024-03-13 CUST001 2024-06-02 CUST065 2024-05-30 CUST180 2024-05-12 CUST035 2024-06-20 CUST117 2024-07-12 CUST005 2024-03-25 CUST041 2024-06-16 CUST102 2024-06-08 CUST146 2024-01-25 CUST131 2024-03-21 CUST138 2024-04-25 CUST141 2024-04-02 CUST136 2024-05-21 CUST066 2024-06-26 CUST149 2024-04-07 CUST054 2024-05-15 CUST119 2024-03-20 CUST040 2024-03-01 CUST031 2024-06-26 CUST025 2024-02-29 CUST167 2024-02-20 CUST097 2024-04-11 CUST143 2024-04-21 CUST076 2024-02-04 CUST015 2024-06-05 CUST030 2024-01-27 CUST009 2024-07-16 CUST168 2024-05-15 CUST082 2024-05-20 CUST086 2024-03-16 CUST092 2024-05-03 CUST125 2024-06-19 CUST156 2024-06-20 CUST153 2024-06-08 CUST116 2024-03-20 CUST074 2024-10-22 CUST078 2024-05-03 CUST079 2024-05-26 CUST158 2024-06-29 CUST193 2024-06-22 CUST200 2024-02-14 CUST064 2024-07-17 CUST069 2024-01-23 CUST126 2024-04-16 CUST120 2024-08-17 CUST071 2024-04-09 CUST190 2024-06-20 CUST075 2024-06-22 CUST199 2024-04-22 CUST060 2024-03-24 CUST145 2024-04-27 CUST058 2024-07-23 CUST135 2024-03-27 CUST173 2024-03-23 CUST073 2024-05-02 CUST152 2024-06-07 CUST183 2024-06-08 CUST166 2024-05-14 CUST112 2024-03-19 CUST045 2024-02-11 CUST070 2024-04-15 CUST008 2024-05-31 CUST154 2024-05-10 CUST194 2024-04-21 CUST067 2024-03-25 CUST104 2024-01-23 CUST091 2024-01-31 CUST046 2024-02-18 CUST189 2024-02-07 CUST088 2024-04-17 CUST188 2024-05-11 CUST123 2024-05-20 CUST107 2024-10-11 CUST033 2024-04-13 CUST108 2024-05-19 CUST110 2024-02-09 CUST068 2024-04-12 CUST027 2024-04-27 CUST122 2024-03-08 CUST047 2024-05-24 CUST038 2024-03-27 CUST162 2024-03-09 CUST014 2024-08-13 CUST191 2024-04-27 CUST129 2024-05-26 CUST016 2024-03-11 CUST161 2024-04-21 CUST140 2024-04-14 CUST022 2024-08-01 CUST048 2024-06-28 CUST160 2024-03-12 CUST101 2024-01-13 CUST178 2024-07-14 CUST164 2024-06-21 CUST039 2024-02-17 CUST044 2024-07-17 CUST072 2024-02-07 CUST052 2024-05-19 CUST159 2024-04-08 CUST124 2024-04-29 CUST011 2024-02-03 CUST195 2024-04-23 CUST197 2024-05-12 CUST055 2024-02-27 CUST134 2024-04-24 CUST128 2024-07-07 CUST098 2024-01-30 CUST096 2024-03-14 CUST130 2024-06-24 CUST007 2024-03-04 CUST144 2024-03-27 CUST049 2024-04-30 CUST182 2024-04-01 CUST099 2024-06-20 CUST174 2024-02-05 CUST148 2024-07-21 CUST002 2024-05-24 CUST083 2024-02-17 CUST118 2024-05-13 CUST179 2024-02-04 CUST151 2024-04-08 CUST062 2024-02-18 CUST133 2024-03-12 CUST004 2024-02-24 CUST111 2024-03-21 CUST198 2024-02-29 CUST084 2024-05-18 CUST165 2024-03-22 CUST024 2024-05-18 CUST021 2024-03-16 CUST121 2024-01-19 CUST170 2024-02-05 CUST061 2024-03-25 CUST184 2024-04-02 CUST003 2024-05-28 CUST100 2024-05-05 CUST089 2024-02-28 CUST106 2024-04-05 CUST019 2024-02-04 CUST150 2024-02-01 CUST177 2024-03-18 CUST147 2024-03-23 CUST176 2024-05-23 CUST109 2024-03-12 CUST087 2024-06-18 CUST187 2024-05-28 CUST132 2024-06-15 CUST037 2024-06-14 CUST137 2024-02-18 CUST023 2024-09-15 CUST113 2024-03-23 CUST127 2024-03-26 CUST186 2024-02-21 CUST053 2024-03-07 CUST080 2024-05-12 CUST115 2024-02-01 CUST163 2024-04-30 CUST077 2024-05-31 CUST093 2024-04-27 CUST018 2024-04-30 CUST105 2024-03-13 CUST036 2024-06-20 CUST032 2024-05-13 CUST020 2024-05-04 CUST155 2024-06-15 CUST063 2024-04-19 CUST095 2024-01-19 CUST050 2024-02-27 CUST042 2024-05-27 CUST171 2024-06-28 CUST056 2024-05-01 CUST114 2024-03-30 CUST090 2024-03-31 CUST081 2024-04-28 CUST051 2024-04-08 CUST029 2024-03-06 CUST172 2024-03-25 CUST010 2024-04-20 CUST043 2024-04-27 CUST196 2024-06-24 CUST192 2024-01-24 CUST017 2024-02-28

# Step 3: Combining the Results and Calculating the Days Between Purchases


In [0]:

# Step 3: Combining first + second purchase and calculate days between
final_df = spark.sql("""
    SELECT
        fp.customer_id,
        fp.first_purchase_date,
        sp.second_purchase_date,
        DATEDIFF(sp.second_purchase_date, fp.first_purchase_date) AS days_between -- calculate # days between purchses
    FROM first_purchases fp
    LEFT JOIN second_purchases sp
        ON fp.customer_id = sp.customer_id --using left join to only include customers with one purchase
""")

# Save as permanent table 
final_df.write.format("delta").mode("overwrite").saveAsTable("workspace.default_cohort_db.cohort_analysis") 


## Verify  Transformation:
## 

In [0]:
%sql
SELECT * FROM workspace.default_cohort_db.cohort_analysis LIMIT 10;

customer_id first_purchase_date second_purchase_date days_between CUST157 2024-01-03 2024-02-04 32 CUST013 2024-04-30 2024-06-23 54 CUST181 2024-01-03 2024-02-11 39 CUST185 2024-03-04 2024-03-26 22 CUST028 2024-06-20 2024-07-28 38 CUST175 2024-05-05 2024-05-30 25 CUST012 2024-02-15 2024-03-04 18 CUST169 2024-02-05 2024-03-14 38 CUST059 2024-02-21 2024-03-12 20 CUST006 2024-01-18 2024-03-31 73

#  Retention Rate by Cohort
# 
Step 1: Create a Query to Compute Retention Rates. 

Write a query that calculates, for each customer (using the raw ecom_orders table), the first purchase date and then (if available) the date of their second order. 
Then, group customers into cohorts by truncating the first purchase date by month and compute the percentage of customers who placed a second order within 1, 2, and 3 months.

In [0]:
##old query using months between
final_ret_df=spark.sql("""
WITH cohorts AS (
    SELECT
        customer_id,
        CAST(DATE_TRUNC('month', first_purchase_date) AS DATE) AS cohort_month,
        second_purchase_date,
        MONTHS_BETWEEN(second_purchase_date, first_purchase_date) AS months_between
    FROM cohort_analysis
    WHERE first_purchase_date IS NOT NULL
),

cohort_aggregates AS (
    SELECT
        cohort_month,
        COUNT(*) AS total_customers,
        COUNT_IF(months_between <= 1) AS second_within_1_month,
        COUNT_IF(months_between <= 2) AS second_within_2_months,
        COUNT_IF(months_between <= 3) AS second_within_3_months
    FROM cohorts
    GROUP BY cohort_month
)



SELECT
    cohort_month,
    total_customers,
    ROUND(100.0 * second_within_1_month / total_customers, 1) AS retention_rate_1_month,
    ROUND(100.0 * second_within_2_months / total_customers, 1) AS retention_rate_2_months,
    ROUND(100.0 * second_within_3_months / total_customers, 1) AS retention_rate_3_months
FROM cohort_aggregates
ORDER BY cohort_month
""")
final_ret_df.display()
# Save as permanent table
final_ret_df.write.format("delta").mode("overwrite").saveAsTable("workspace.default_cohort_db.retention_rate_by_cohort")


cohort_month total_customers retention_rate_1_month retention_rate_2_months retention_rate_3_months 2024-01-01 66 37.9 60.6 77.3 2024-02-01 48 35.4 66.7 79.2 2024-03-01 33 33.3 72.7 87.9 2024-04-01 27 51.9 88.9 92.6 2024-05-01 16 56.3 93.8 100.0 2024-06-01 10 30.0 100.0 100.0

Databricks visualization. Run in Databricks to view.

In [0]:
final_ret_df = spark.sql("""
WITH first_and_second_orders AS (
    SELECT
        customer_id,
        order_date,
        --Assigns a sequential number to each order per customer based on the earliest order first.
        --This is used to identify the first and second orders for each customer.
        ROW_NUMBER() OVER (PARTITION BY customer_id ORDER BY order_date) AS order_rank 
    FROM ecom_orders
),

customer_orders AS (
    SELECT
        customer_id,
        MIN(CASE WHEN order_rank = 1 THEN order_date END) AS first_purchase_date,
        MIN(CASE WHEN order_rank = 2 THEN order_date END) AS second_purchase_date
    FROM first_and_second_orders
    GROUP BY customer_id
),

cohorts AS (
    SELECT
        customer_id,
        CAST(DATE_TRUNC('month', first_purchase_date) AS DATE) AS cohort_month,
        first_purchase_date,
        second_purchase_date,
        DATEDIFF(second_purchase_date, first_purchase_date) AS days_between
    FROM customer_orders
    WHERE first_purchase_date IS NOT NULL
)

SELECT
    cohort_month,
    COUNT(*) AS total_customers,
    SUM(CASE WHEN second_purchase_date IS NOT NULL AND days_between <= 30 THEN 1 ELSE 0 END) AS second_within_1_month,
    SUM(CASE WHEN second_purchase_date IS NOT NULL AND days_between <= 60 THEN 1 ELSE 0 END) AS second_within_2_months,
    SUM(CASE WHEN second_purchase_date IS NOT NULL AND days_between <= 90 THEN 1 ELSE 0 END) AS second_within_3_months,
    ROUND(100.0 * SUM(CASE WHEN second_purchase_date IS NOT NULL AND days_between <= 30 THEN 1 ELSE 0 END) / COUNT(*), 1) AS retention_rate_1_month,
    ROUND(100.0 * SUM(CASE WHEN second_purchase_date IS NOT NULL AND days_between <= 60 THEN 1 ELSE 0 END) / COUNT(*), 1) AS retention_rate_2_months,
    ROUND(100.0 * SUM(CASE WHEN second_purchase_date IS NOT NULL AND days_between <= 90 THEN 1 ELSE 0 END) / COUNT(*), 1) AS retention_rate_3_months
FROM cohorts
GROUP BY cohort_month
ORDER BY cohort_month
;


""")

# Show in notebook
final_ret_df.display()

# Save as permanent table

spark.sql("DROP TABLE IF EXISTS workspace.default_cohort_db.retention_rate_by_cohort")
final_ret_df.write.format("delta").saveAsTable("workspace.default_cohort_db.retention_rate_by_cohort")


cohort_month total_customers second_within_1_month second_within_2_months second_within_3_months retention_rate_1_month retention_rate_2_months retention_rate_3_months 2024-01-01 66 26 41 51 39.4 62.1 77.3 2024-02-01 48 17 32 38 35.4 66.7 79.2 2024-03-01 33 10 24 29 30.3 72.7 87.9 2024-04-01 27 14 24 25 51.9 88.9 92.6 2024-05-01 16 8 15 16 50.0 93.8 100.0 2024-06-01 10 3 10 10 30.0 100.0 100.0

Databricks visualization. Run in Databricks to view.

Databricks visualization. Run in Databricks to view.

In [0]:
%sql
SELECT customer_id, 
COUNT(DISTINCT order_id) AS num_orders
FROM ecom_orders
GROUP BY customer_id
HAVING num_orders = 1


customer_id num_orders CUST094 1 CUST026 1

In [0]:
##Check for customers with only 1 order

spark.sql("""
SELECT
    COUNT(*) AS one_order_customers
FROM (
    SELECT customer_id, 
    COUNT(DISTINCT order_id) AS num_orders
    FROM ecom_orders
    GROUP BY customer_id
    HAVING num_orders = 1
)
""").display()


one_order_customers 2

In [0]:

## Check for duplicate order_ids

spark.sql("""
SELECT
    order_id,
    COUNT(*) AS occurrences
FROM ecom_orders
GROUP BY order_id
HAVING COUNT(*) > 1
ORDER BY occurrences DESC
""").display()



order_id occurrences

In [0]:
%sql
SELECT
  customer_id,
  COUNT(DISTINCT order_id) AS order_count,
  MIN(order_date) AS first_order
FROM ecom_orders
GROUP BY customer_id
HAVING order_count = 1

customer_id order_count first_order CUST094 1 2024-04-03 CUST026 1 2024-02-14

In [0]:
%sql
SELECT
  DATE_TRUNC('month', DATE(first_order)) AS cohort_month,
  COUNT(*)
FROM (
 SELECT
  customer_id,
  COUNT(DISTINCT order_id) AS order_count,
  MIN(order_date) AS first_order
FROM ecom_orders
GROUP BY customer_id
HAVING order_count = 1
)
GROUP BY 1


cohort_month COUNT(*) 2024-02-01T00:00:00.000Z 1 2024-04-01T00:00:00.000Z 1

In [0]:
%sql
SELECT order_id, COUNT(*) 
FROM ecom_orders
GROUP BY order_id
HAVING COUNT(*) > 1
LIMIT 10


order_id COUNT(*)

In [0]:
%sql
WITH customer_orders AS (
  SELECT
    customer_id,
    DATE_TRUNC('month', MIN(order_date)) AS cohort_month
  FROM ecom_orders
  GROUP BY customer_id
)
SELECT cohort_month, COUNT(*) FROM customer_orders GROUP BY 1 ORDER BY 1



cohort_month COUNT(*) 2024-01-01T00:00:00.000Z 66 2024-02-01T00:00:00.000Z 48 2024-03-01T00:00:00.000Z 33 2024-04-01T00:00:00.000Z 27 2024-05-01T00:00:00.000Z 16 2024-06-01T00:00:00.000Z 10

# Repeat Purchase Rate by Cohort

Step 1: Create a Query to Compute Repeat Purchase Rates. 

Write a query that calculates, for each customer, the first purchase date and total orders.  

Then, group customers into monthly cohorts and compute the percentage of customers who placed at least a 2nd, 3rd, and 4th order.

In [0]:
repeat_purchase_df = spark.sql("""
WITH customer_orders AS (
    SELECT
        customer_id,
        CAST(DATE_TRUNC('month', MIN(order_date)) AS DATE) AS cohort_month,
        COUNT(DISTINCT order_id) AS total_orders
    FROM ecom_orders
    GROUP BY customer_id
),

cohort_stats AS (
    SELECT
        cohort_month,
        COUNT(*) AS total_customers,
        COUNT_IF(total_orders >= 2) AS repeat_2_orders,
        COUNT_IF(total_orders >= 3) AS repeat_3_orders,
        COUNT_IF(total_orders >= 4) AS repeat_4_orders
    FROM customer_orders
    GROUP BY cohort_month
)

SELECT
    cohort_month,
    total_customers,
    ROUND(100.0 * repeat_2_orders / total_customers, 1) AS repeat_rate_2_orders,
    ROUND(100.0 * repeat_3_orders / total_customers, 1) AS repeat_rate_3_orders,
    ROUND(100.0 * repeat_4_orders / total_customers, 1) AS repeat_rate_4_orders
FROM cohort_stats
ORDER BY cohort_month
""")

repeat_purchase_df.display()
# Save as permanent table
repeat_purchase_df.write.format("delta").mode("overwrite").saveAsTable("workspace.default_cohort_db.repeat_purchase_rates_by_cohort")


cohort_month total_customers repeat_rate_2_orders repeat_rate_3_orders repeat_rate_4_orders 2024-01-01 66 100.0 98.5 83.3 2024-02-01 48 97.9 95.8 81.3 2024-03-01 33 100.0 93.9 81.8 2024-04-01 27 96.3 85.2 77.8 2024-05-01 16 100.0 93.8 68.8 2024-06-01 10 100.0 70.0 60.0

Databricks visualization. Run in Databricks to view.

# **Cohort Size by Month**
Step 1: Create a Query to Compute Cohort Size. 

Write a query to count the number of new customers (i.e., first orders) acquired in each month.

In [0]:
# Compute Cohort Size by First Purchase Month
cohort_size_by_month_df = spark.sql("""
WITH first_orders AS (
    SELECT
        customer_id,
        MIN(order_date) AS first_purchase_date
    FROM ecom_orders
    GROUP BY customer_id
)

SELECT
    Cast(DATE_TRUNC('month', first_purchase_date) as Date) AS cohort_month,
    COUNT(*) AS new_customers
FROM first_orders
GROUP BY DATE_TRUNC('month', first_purchase_date)
ORDER BY cohort_month
""")
display(cohort_size_by_month_df)
# Save as permanent table
cohort_size_by_month_df.write.format("delta").mode("overwrite").saveAsTable("workspace.default_cohort_db.cohort_size_by_month")


cohort_month new_customers 2024-01-01 66 2024-02-01 48 2024-03-01 33 2024-04-01 27 2024-05-01 16 2024-06-01 10

Databricks visualization. Run in Databricks to view.

In [0]:
repeat_purchase_df = spark.sql("""
-- Step 1: Deduplicate orders at customer + order_id level
WITH deduped_orders AS (
    SELECT
        customer_id,
        order_id,
        MIN(order_date) AS order_date  -- in case of duplicates, pick earliest
    FROM ecom_orders
    GROUP BY customer_id, order_id
),

-- Step 2: Get first order month + count total orders
customer_orders AS (
    SELECT
        customer_id,
        CAST(DATE_TRUNC('month', MIN(order_date)) AS DATE) AS cohort_month,
        COUNT(*) AS total_orders  -- no need for DISTINCT anymore
    FROM deduped_orders
    GROUP BY customer_id
),

-- Step 3: Aggregate repeat purchase stats by cohort
cohort_stats AS (
    SELECT
        cohort_month,
        COUNT(*) AS total_customers,
        COUNT_IF(total_orders >= 2) AS repeat_2_orders,
        COUNT_IF(total_orders >= 3) AS repeat_3_orders,
        COUNT_IF(total_orders >= 4) AS repeat_4_orders
    FROM customer_orders
    GROUP BY cohort_month
)

-- Step 4: Final results
SELECT
    cohort_month,
    total_customers,
    ROUND(100.0 * repeat_2_orders / total_customers, 1) AS repeat_rate_2_orders,
    ROUND(100.0 * repeat_3_orders / total_customers, 1) AS repeat_rate_3_orders,
    ROUND(100.0 * repeat_4_orders / total_customers, 1) AS repeat_rate_4_orders
FROM cohort_stats
ORDER BY cohort_month
""")

# Display results
repeat_purchase_df.display()

# Save as a Delta table
repeat_purchase_df.write.format("delta").mode("overwrite").saveAsTable("workspace.default_cohort_db.repeat_purchase_rates_by_cohort_deduped")


cohort_month total_customers repeat_rate_2_orders repeat_rate_3_orders repeat_rate_4_orders 2024-01-01 66 100.0 98.5 83.3 2024-02-01 48 97.9 95.8 81.3 2024-03-01 33 100.0 93.9 81.8 2024-04-01 27 96.3 85.2 77.8 2024-05-01 16 100.0 93.8 68.8 2024-06-01 10 100.0 70.0 60.0

In [0]:
final_ret_df=spark.sql("""
WITH cohorts AS (
    SELECT
        customer_id,
        CAST(DATE_TRUNC('month', first_purchase_date) AS DATE) AS cohort_month,
        second_purchase_date,
        MONTHS_BETWEEN(second_purchase_date, first_purchase_date) AS months_between
    FROM workspace.default_cohort_db.cohort_analysis
    WHERE first_purchase_date IS NOT NULL
),

cohort_aggregates AS (
    SELECT
        cohort_month,
        COUNT(*) AS total_customers,
        COUNT_IF(months_between <= 1) AS second_within_1_month,
        COUNT_IF(months_between <= 2) AS second_within_2_months,
        COUNT_IF(months_between <= 3) AS second_within_3_months
    FROM cohorts
    GROUP BY cohort_month
)

SELECT
    cohort_month,
    total_customers,
    ROUND(100.0 * second_within_1_month / total_customers, 1) AS retention_rate_1_month,
    ROUND(100.0 * second_within_2_months / total_customers, 1) AS retention_rate_2_months,
    ROUND(100.0 * second_within_3_months / total_customers, 1) AS retention_rate_3_months
FROM cohort_aggregates
ORDER BY cohort_month
""")
final_ret_df.display()
# Save as permanent table
##final_ret_df.write.format("delta").mode("overwrite").saveAsTable("workspace.demo.retention_rate_by_cohort")


cohort_month total_customers retention_rate_1_month retention_rate_2_months retention_rate_3_months 2024-01-01 66 37.9 60.6 77.3 2024-02-01 48 35.4 66.7 79.2 2024-03-01 33 33.3 72.7 87.9 2024-04-01 27 51.9 88.9 92.6 2024-05-01 16 56.3 93.8 100.0 2024-06-01 10 30.0 100.0 100.0

**Observations**. 

**1. Declining acquisition**. 

Fewer new customers each month, from 66 in January down to 10 in June, and no new customers until the end of the year

Maybe a drop in acquisition, which could be seasonal, strategic, or an issue (e.g. reduced marketing spend or fewer leads).

**2. Improving retention**. 
Even though acquisition is shrinking, retention is getting better:

**Jan 2024:**   
Only 39.4% of customers returned within 1 month.

**Apr–May 2024**: Over 50% return within 1 month.

In **May/June 2024**, almost all customers return within 2–3 months.

**3. March vs. May cohorts:**. 

**March** had low early retention (30.3%) but very high later retention (87.9%)

**May** had smaller volume (16 customers) but perfect 3-month retention (100%)




**Interpretation and Conclusions**. 

- Quality of customers is improving, even if volume is dropping.

- Retention is climbing steadily, especially long-term.
 
- If the goal is growth, the shrinking cohort sizes could become a problem, even if LTV per customer is rising.
 
- Segment further by acquisition channel, product, or location to see what’s driving strong retention in Apr–Jun.